docs/split-inputs.md

```diff
@@ -0,0 +1,140 @@
+a single fuzzer-generated input into several.
+
+# Examples
+Splitting a fuzzer-generated input into several independent inputs is required
+very often. Some examples:
+
+* Fuzzing a regular expression library requires
+   * The regular expression (RE)
+   * Flags for RE compilation and matching
+   * A string to search the RE in
+
+* Fuzzing an audio/video format decoder often requires
+  * Decoding flags
+  * Several frames
+
+* Fuzzing a XSLT or CSS library requires
+  * The stylesheet input
+  * The XML/HTML input
+
+* Fuzzing a font-rendering library requires
+  * The font file
+  * The text to render
+  * The rendering flags
+
+* Fuzzing a database library may require
+  * The query text
+  * The database state
+
+TODO: more examples?
+
+# Common Data Format
+
+When trying to split the fuzzer-generated input into several,
+the first question one needs to ask is whether the input format is common,
+i.e. is it used or processed by other libraries, APIs, of fuzz targets.
+
+If the data format is common (e.g. a widely used media format or network packet
+format) then it is highly desirable for a fuzz target to consume exactly this
+data format, and not some custom modification.
+This way it will be easier to procure a seed corpus for this fuzz target
+and to use the generated corpus to test/fuzz other targets.
+
+## Multiple options
+If the data format may be processed by a fuzz target in a small number of different ways,
+it is often the best approach to split the fuzz target into several ones,
+each processing the input in exactly one way.
+
+Make sure to cross-pollinate the corpora between these targets.
+OSS-Fuzz does that automatically.
+
+## Embedding / Comments
+When a fuzz target for a common data format requires some flags, options, or
+additional auxiliary sub-input(s), it is sometimes possible to embed the extra input
+inside a custom section or a comment of the main data format.
+
+Examples:
+* PNG allows custom "chunks", and so a fuzz target for a PNG decoder can
+  hide the flags used during PNG processing
+  in a separate PNG chunk, e.g. `fUZz`
+  ([example](https://github.com/google/oss-fuzz/blob/master/projects/libpng-proto/libpng_transforms_fuzzer.cc)).
+* When fuzzing C/C++/Java/JavaScript inputs one may hide a sub-input in a single-line `//` comment. TODO: example?
+
+
+## Hash
+When only one small fixed-size sub-input is required (such as flags / options),
+the fuzz target may compute a hash function on the full input and use it as the flag bits.
+This option is very easy to implement, but it's applicability is limited to
+relatively simple cases. The major problem is that a small local mutation of the input
+leads to a large change in the sub-input, which often makes fuzzing less
+efficient. Try this approach if the flags are individual bits and the input
+type allows some bit flips in the inputs (e.g. a plain text).
+
+TODO: example.
+
+# Custom Serialization Format
```

Comment:
Maybe move this to the end as well, this seems to be pretty rare case to me.

Reply:
This is a section that contains subsections on everything else.

Reply:
Ah, stupid me!